File: example/rnn/old/lstm_bucketing.py
# pylint: disable=C0111,too-many-arguments,too-many-instance-attributes,too-many-locals,redefined-outer-name,fixme
# pylint: disable=superfluous-parens, no-member, invalid-name
import sys
sys.path.insert(0, "../../python")
import numpy as np
import mxnet as mx

from lstm import lstm_unroll
from bucket_io import BucketSentenceIter, default_build_vocab

def Perplexity(label, pred):
    label = label.T.reshape((-1,))
    loss = 0.
    for i in range(pred.shape[0]):
        loss += -np.log(max(1e-10, pred[i][int(label[i])]))
    return np.exp(loss / label.size)

if __name__ == '__main__':
    N = 8
    batch_size = 32*N
    #buckets = [10, 20, 30, 40, 50, 60]
    buckets = [32]
    #buckets = []
    num_hidden = 200
    num_embed = 200
    num_lstm_layer = 2

    num_epoch = 25
    learning_rate = 0.01
    momentum = 0.0

    # dummy data is used to test speed without IO
    dummy_data = False

    contexts = [mx.context.gpu(i) for i in range(N)]

    vocab = default_build_vocab("./data/ptb.train.txt")

    def sym_gen(seq_len):
        return lstm_unroll(num_lstm_layer, seq_len, len(vocab),
                           num_hidden=num_hidden, num_embed=num_embed,
                           num_label=len(vocab))

    init_c = [('l%d_init_c'%l, (batch_size, num_hidden)) for l in range(num_lstm_layer)]
    init_h = [('l%d_init_h'%l, (batch_size, num_hidden)) for l in range(num_lstm_layer)]
    init_states = init_c + init_h

    data_train = BucketSentenceIter("./data/ptb.train.txt", vocab,
                                    buckets, batch_size, init_states)
    data_val = BucketSentenceIter("./data/ptb.valid.txt", vocab,
                                  buckets, batch_size, init_states)

    if dummy_data:
        data_train = DummyIter(data_train)
        data_val = DummyIter(data_val)

    if len(buckets) == 1:
        # only 1 bucket, disable bucketing
        symbol = sym_gen(buckets[0])
    else:
        symbol = sym_gen

    model = mx.model.FeedForward(ctx=contexts,
                                 symbol=symbol,
                                 num_epoch=num_epoch,
                                 learning_rate=learning_rate,
                                 momentum=momentum,
                                 wd=0.00001,
                                 initializer=mx.init.Xavier(factor_type="in", magnitude=2.34))

    import logging
    head = '%(asctime)-15s %(message)s'
    logging.basicConfig(level=logging.DEBUG, format=head)

    model.fit(X=data_train, eval_data=data_val, kvstore='device',
              eval_metric = mx.metric.np(Perplexity),
              batch_end_callback=mx.callback.Speedometer(batch_size, 50),)

File: scala-package/examples/src/main/scala/ml/dmlc/mxnetexamples/rnn/LstmBucketing.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package ml.dmlc.mxnetexamples.rnn

import ml.dmlc.mxnet.Callback.Speedometer
import ml.dmlc.mxnet._
import BucketIo.BucketSentenceIter
import ml.dmlc.mxnet.optimizer.SGD
import org.kohsuke.args4j.{CmdLineParser, Option}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.JavaConverters._
import ml.dmlc.mxnet.module.BucketingModule
import ml.dmlc.mxnet.module.FitParams

/**
 * Bucketing LSTM examples
 * @author Yizhi Liu
 */
class LstmBucketing {
  @Option(name = "--data-train", usage = "training set")
  private val dataTrain: String = "example/rnn/ptb.train.txt"
  @Option(name = "--data-val", usage = "validation set")
  private val dataVal: String = "example/rnn/ptb.valid.txt"
  @Option(name = "--num-epoch", usage = "the number of training epoch")
  private val numEpoch: Int = 5
  @Option(name = "--gpus", usage = "the gpus will be used, e.g. '0,1,2,3'")
  private val gpus: String = null
  @Option(name = "--cpus", usage = "the cpus will be used, e.g. '0,1,2,3'")
  private val cpus: String = null
  @Option(name = "--save-model-path", usage = "the model saving path")
  private val saveModelPath: String = "model/lstm"
}

object LstmBucketing {
  private val logger: Logger = LoggerFactory.getLogger(classOf[LstmBucketing])

  def perplexity(label: NDArray, pred: NDArray): Float = {
    pred.waitToRead()
    val labelArr = label.T.toArray.map(_.toInt)
    var loss = .0
    (0 until pred.shape(0)).foreach(i =>
      loss -= Math.log(Math.max(1e-10f, pred.slice(i).toArray(labelArr(i))))
    )
    Math.exp(loss / labelArr.length).toFloat
  }

  def main(args: Array[String]): Unit = {
    val inst = new LstmBucketing
    val parser: CmdLineParser = new CmdLineParser(inst)
    try {
      parser.parseArgument(args.toList.asJava)
      val contexts =
        if (inst.gpus != null) inst.gpus.split(',').map(id => Context.gpu(id.trim.toInt))
        else if (inst.cpus != null) inst.cpus.split(',').map(id => Context.cpu(id.trim.toInt))
        else Array(Context.cpu(0))

      val batchSize = 32
      val buckets = Array(10, 20, 30, 40, 50, 60)
      val numHidden = 200
      val numEmbed = 200
      val numLstmLayer = 2

      logger.info("Building vocab ...")
      val vocab = BucketIo.defaultBuildVocab(inst.dataTrain)

      def BucketSymGen(key: AnyRef):
        (Symbol, IndexedSeq[String], IndexedSeq[String]) = {
        val seqLen = key.asInstanceOf[Int]
        val sym = Lstm.lstmUnroll(numLstmLayer, seqLen, vocab.size,
          numHidden = numHidden, numEmbed = numEmbed, numLabel = vocab.size)
        (sym, IndexedSeq("data"), IndexedSeq("softmax_label"))
      }

      val initC = (0 until numLstmLayer).map(l =>
        (s"l${l}_init_c_beta", (batchSize, numHidden))
      )
      val initH = (0 until numLstmLayer).map(l =>
        (s"l${l}_init_h_beta", (batchSize, numHidden))
      )
      val initStates = initC ++ initH

      val dataTrain = new BucketSentenceIter(inst.dataTrain, vocab,
        buckets, batchSize, initStates)
      val dataVal = new BucketSentenceIter(inst.dataVal, vocab,
        buckets, batchSize, initStates)

      val model = new BucketingModule(
        symGen = BucketSymGen,
        defaultBucketKey = dataTrain.defaultBucketKey,
        contexts = contexts)

      val fitParams = new FitParams()
      fitParams.setEvalMetric(
        new CustomMetric(perplexity, name = "perplexity"))
      fitParams.setKVStore("device")
      fitParams.setOptimizer(
        new SGD(learningRate = 0.01f, momentum = 0f, wd = 0.00001f))
      fitParams.setInitializer(new Xavier(factorType = "in", magnitude = 2.34f))
      fitParams.setBatchEndCallback(new Speedometer(batchSize, 50))

      logger.info("Start training ...")
      model.fit(
        trainData = dataTrain,
        evalData = Some(dataVal),
        numEpoch = inst.numEpoch, fitParams)
      logger.info("Finished training...")
    } catch {
      case ex: Exception =>
        logger.error(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
    }
  }
}
